### DISCUSSION TITLE
Azure: Seq Server Appears to Stop Receiving Events after AppService Reboots

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [datalust/seq-tickets#1651] [Help] [answered]
Hi Team Datalust,

Seq has been working amazingly for me ever since I took the pains (and more pains 😅) of setting up a Linux container instance and running Seq there.

Today, however, I ran into a problem and am still attempting to figure out what happened.

At some point at about 1PM ET, Seq reset not once, but twice.  After it reset, the diagnostics from my entire system stopped, and I was left dead in the water from a logging perspective.

Here is what I see in the Container Instance Log in Azure (not sure if I should blur out the IDs but being safer more than sorrier):

![image](https://user-images.githubusercontent.com/2931376/188511240-8231aaa9-37f7-4626-a918-8c5b373f5214.png)

(Outside of this massively disruptive incident, 2 weeks uptime is pretty respectable)

After poking around some, it appears that Seq was OK after the reboot, but because it went offline, all my Seq/Serilog clients in all my services (8 in total) were in a failed state _and remained in a failed state_.  After doing a full redeployment (causing an IIS reset) all my services are back online and operating again.

There are two issues here.

1. Why did Seq reset?
2. How do I configure my Seq/Serilog clients so that they do not remain in a failed state and an AppService/IIS reset is not required?

Thank you for any assistance you can provide.

**Comment by nblumhardt** [comment]
Hey Mike! My guess is that Seq would have been restarted by the hosting infrastructure e.g. because of maintenance (like a VM restart?) but sending the diagnostic report from Seq (Settings > Diagnostics) to support@ would give us a chance to spot any problems in Seq itself.

Serilog on the client uses exponential backoff to avoid swamping a struggling/rebooting server; this means it might take up to 10 minutes for logs to start flowing again after a restart (Serilog will generally buffer events during some or all of this time). Restarting the clients will reset the connection schedule, but leaving the client alone will also usually sort things out. How long after Seq came back up were the clients disrupted for?

Cheers,
Nick

**Reply by Mike-E-angelo** [reply]
> but sending the diagnostic report from Seq (Settings > Diagnostics) to support@ would give us a chance to spot any problems in Seq itself.

Done 👍

> How long after Seq came back up were the clients disrupted for?

The entire time until I did a system-wide IIS reset. :(  There were 0 new logs during that time.  That is why I thought there was something wrong with Seq again but it was really the clients that were not sending data, for whatever reason.

**Comment by nblumhardt** [comment]
Hi Mike, thanks for sending the report!

It looks like ACI just shut the container down cleanly in both of these instances:

```
{"@t":"2022-09-05T14:14:27.9870172Z","@mt":"Termination signal received"}
...
{"@t":"2022-09-05T16:55:04.5791432Z","@mt":"Termination signal received"}
```

When you say "the entire time" how long do you mean? Your Serilog/Seq sink configuration + this info would help determine whether it's exponential backoff kicking in on the client.

It'd also be great to know if you have any kind of app gateway/reverse proxy in between the clients and Seq.

Hope this helps,
Nick

**Reply by Mike-E-angelo** [reply]
OK thank you for the investigation and patience here in dealing with my impatience, @nblumhardt.  Again. 😁  Is there a way to have ACI not shut down in the middle of the day?  I am also unclear why it stopped not once but twice.  My guess is that Seq survived the first one, but the second one probably did it in?  Although the times do not line up for my theory there (more below).

Pardon further the unclarity on my side regarding "the entire time."  What I meant by this was after the restart I stopped receiving messages at 12:53pm ET.  No messages were received until I did a system redeployment and only then were messages emitted from Seq.  In total it was about 3.5 hours:

![image](https://user-images.githubusercontent.com/2931376/188812354-e193b3f5-5e16-4eda-a18d-0e2159abfa45.png)

Note that no proxy/gateway is used.

My configuration for Serilog is pretty vanilla, but maybe you can spot something that would keep messages flowing to my Seq server.  Every client application uses the following:

```
	"Logging": {
		"LogLevel": {
			"Default": "Information",
			"Microsoft": "Warning",
			"Microsoft.Hosting.Lifetime": "Information"
		}
	},
	"Serilog": {
		"Using": [ "DragonSpark.Diagnostics" ],
		"LevelSwitches": { "$controlSwitch": "Information" },
		"MinimumLevel": { "ControlledBy": "$controlSwitch" },
		"Enrich": [
			"WithClientIp",
			"WithClientAgent",
			"WithDemystifiedStackTraces",
			"WithExceptionStackTraceHash",
			"WithEnvironmentName",
			"WithEnvironmentUserName",
			"FromLogContext",
			"WithProcessId",
			"WithProcessName",
			"WithMemoryUsage",
			"WithThreadId",
			"WithThreadName",
			"WithCorrelationId",
			"WithEnvironmentUserName",
			"EnvironmentLoggerConfigurationExtensions.WithMachineName",
			"WithFrameworkEnrichers"
		],
    		"WriteTo": [
			{
				"Name": "Seq",
				"Args": {
					"serverUrl": "https://<URL>",
					"controlLevelSwitch": "$controlSwitch"
				}
			}
		]
	},
```

**Reply by nblumhardt** [reply]
Thanks for the follow-up, Mike!

The Azure docs [state that](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-troubleshooting#container-had-an-isolated-restart-without-explicit-user-input)  _customers may experience restarts initiated by the ACI infrastructure due to maintenance events_, which could have been the case here.

If there is any more info to unearth, though, you might be able to spot something in `az container logs` - worth another check if you haven't grepped through them already.

Your Serilog configuration looks fine :thinking:  ....

I'm guessing you're running your client code on .NET (core) 5 or 6? I have a recollection of `HttpClient` (which _Serilog.Sinks.Seq_ uses) having trouble getting its act back together in some cases when DNS changes; one possible cause of the long outage is that when the container restarted, it may not have popped back up with the same IP address; the clients would need to query DNS again to get the new address, if so.

The [docs for `HttpClient`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.pooledconnectionlifetime?view=net-6.0#system-net-http-socketshttphandler-pooledconnectionlifetime) suggest that the default pooled connection lifetime (infinite) is only suitable for connecting via IP address:

> If the connection endpoint is not the Domain name but the IP address, the value can be [InfiniteTimeSpan](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timeout.infinitetimespan?view=net-6.0).

so we could be on the right track, here.

Overriding it looks something like:

```csharp
    .WriteTo.Seq("https://<URL>", controlLevelSwitch: leveSwitch, messageHandler: new SocketsHttpHandler
    {
        PooledConnectionIdleTimeout = TimeSpan.FromMinutes(1),
        PooledConnectionLifetime = TimeSpan.FromMinutes(1),
    });
```

Let me know if you want to give this a try and need any help configuring it.

https://docs.microsoft.com/en-us/azure/container-instances/container-instances-application-gateway has instructions for giving your container a static IP address, which might be a good option long-term.

Hope this helps,
Nick

**Reply by Mike-E-angelo** [reply]
That's good enough for me, @nblumhardt.  It is very valuable information and I feel you landed on the correct diagnosis here.  I have another ticket outstanding with Azure about my said AppServices which also reset several times a few weeks ago.  I will throw this incident in there to be sure everything is accounted for.

One last question regarding the suggested HTTP configuration: is that possible via JSON configuration?

**Reply by nblumhardt** [reply]
Hi Mike; no, not via JSON configuration. I think with some `#if` conditionals, it might be an option for us to enable it by default in `WriteTo.Seq()` on platforms supported by `SocketsHttpHandler`, I'll drop a ticket over in that repository to investigate.

**Reply by Mike-E-angelo** [reply]
Thank you @nblumhardt!  If you have a link I would greatly appreciate it so I can follow along.  Thank you again for your assistance/patience here. 🙏

**Reply by nblumhardt** [reply]
Hi Mike; https://github.com/datalust/serilog-sinks-seq/issues/178 is the ticket :+1:

**Reply by Mike-E-angelo** [reply]
Hi @nblumhardt it appears I ran into this issue again, where logs stopped working for all my AppServices (7 of them) connected to my diagnostics (Seq) Container Instance.  Azure infrastructure restarted not just once or twice yesterday but THREE full times, so that appears to be a part of what triggers this.

FWIW here's the version of my Seq server: 
![image](https://user-images.githubusercontent.com/2931376/197222236-0d37a50a-beec-4272-8e9a-ed3b42e605e2.png)

It looks like restarting my diagnostics (Seq) server followed by my AppServices is what gets logging working again.  However, from what I understand, that is not supposed to happen, correct?  I was wanting to check in on the expected version for this fix to be in and also make sure I understand the expected remedy with the correct version of Seq server and also for the Serilog sink.  I did do a NuGet update a few days ago so I should be on latest RTM versions for everything.

If the fix is still in dev/preview then that would be why this is still occurring in my environment.  Just doing a quick sanity check here to be sure. :)

Thank you again for your assistance and any remaining clarity you can provide.

**Reply by Mike-E-angelo** [reply]
I'd also like to say that one of those Azure shutdowns occurred at 3:45PM ET which is basically prime time for me.  Good times. 🥳  You'd think Azure would know by now the best time to rain disruption upon their customers? 🤔

**Reply by nblumhardt** [reply]
Hi Mike - the fix for the pooled connection lifetime issue is client-side, in Serilog.Sinks.Seq v5.2.0 - which version are your apps running? Thanks!

**Reply by Mike-E-angelo** [reply]
Hmm... I've been on 5.2.1 since October 9th:
https://github.com/DragonSpark/Framework/blob/master/DragonSpark.Diagnostics/DragonSpark.Diagnostics.csproj#L12

With a full deployment to all my AppServices on October 15th.

So that would suggest maybe another issue may be at play here then?

This is what I experience:
- No logging
- Reboot Seq Container Instance
- Reboot AppServices
- Logging resumes

Logging does not resume from when it stopped but from the current point in time.  So in effect, I lose all messages created during this window of time.

**Reply by nblumhardt** [reply]
Thanks for the extra info, Mike; ... 🤔 interesting ... what happens if you just reboot your services, and not also the Seq container instance?

On the client-side, we might see some more details if you direct Serilog's `SelfLog` to `Console.Error`, and then collect the container logs from the applications when the problem shows up.

It might also be a good time for us to check through another diagnostic report from the Seq instance. The reports contain Seq's internal log messages; do you see much in there during the time you experienced an outage? Feel free to blast the report through our way if you like us to check it out. Thanks!

**Reply by Mike-E-angelo** [reply]
> what happens if you just reboot your services, and not also the Seq container instance?

This appears to be what starts this issue.  AppServices reboot and logging events stop across all instances.  It is not until the Container Instance is rebooted (and it appears all AppServices also reboot) that logging starts to occur again.

> and then collect the container logs from the applications when the problem shows up.

When you say `container logs` note that the clients here (where `Serilog.Sinks.Seq` is present) are AppService instances.  I say this as I am currently unclear where container logs are for AppService instances. :)

> Feel free to blast the report through our way if you like us to check it out

Done 👍

**Reply by nblumhardt** [reply]
Thanks for that @Mike-E-angelo.

I can see the five restarts in succession:

![image](https://user-images.githubusercontent.com/342712/197643035-aabc535a-d3e6-4387-8db4-1f5ac8435756.png)

Each time there's a termination signal received, followed shortly by the startup sequence, all within a very narrow window (all within 30 seconds) - all nice and clean, nothing unusual on the Seq side that I can see 🤔 

If I'm understanding you correctly, you can still reach Seq's UI during the gaps; if you are at a machine with `seqcli` or `docker`, and this happens again, it will be useful to try running:

```
seqcli log -m "Test" -s <public Seq URL> -a <API key>
```

This will test the ingestion endpoint and help to show up any reasons that the inbound requests are getting bounced.

On the App Service side, turning on diagnostic logging is buried in the docs a bit - here's the entrypoint:

https://learn.microsoft.com/en-us/azure/app-service/troubleshoot-diagnostic-logs#enable-application-logging-linuxcontainer

The logs can then be retrieved via the `az` command and in the Azure Portal, as far as I can tell.

Hope this helps,
Nick

**Reply by Mike-E-angelo** [reply]
Great, thank you @nblumhardt your time and attention here are appreciated.  My level of anxiety around this issue has gone from a 8-9 to a 1-2 as there is a workaround now and its impact is reduced.  Even if it's disruptive, being able to use the turn-off-turn-on method as described above is good enough for me while we track this down.

The only catch here is realizing that logging has stopped and catching it in time. 😅  Are you able to verify on your end that events stopped shortly after those restarts after 6:03am on October 21st?

In any case I will be mindful going forward of restarts and checking diligently if logging is intact and attempt the above if it stops.

Please note that there appears to be confusion on my Seq server... it is not an AppService but a Container Instance.  the logging instructions you provided above is for linux AppService which I do not have. :)  

So:

- Azure Container Instance (Linux): Seq server install
- AppServices (Windows): 7 client instances connected to the above

That stated I am not sure if there is a CLI for Container Instances?  I tried using the `Connect` tab and it is not cooperating: 
![image](https://user-images.githubusercontent.com/2931376/197722781-7a1f8b9f-91ae-40d7-a317-c35fc21dbd34.png)

Any further assistance to have this locked and loaded the next time this occurs would be appreciated.

**Reply by nblumhardt** [reply]
Hi @Mike-E-angelo - thanks for your reply. Sorry I think I was unclear about a couple of things :-)

The `seqcli` command should be run on another machine entirely, ideally your development machine or some similar place where you can also reach the Seq UI. The `seqcli` binary is included in the Seq Windows installer, or you can grab platform-specific binaries from https://github.com/datalust/seqcli/releases.

I did mean App Services when I was suggesting collecting Serilog's `SelfLog` from `Console.Error` - I might have misunderstood how you're deploying to App Service, though (I'd assumed you were using containers, there). If you're deploying non-containerized apps to App Service then I think the instructions will be different. The main goal is to pipe Serilog's `SelfLog` to somewhere you can collect it on your application side. You might need to have a dig in the Azure docs to find details, depending on how you have things set up.

Hope this helps,
Nick

**Reply by Mike-E-angelo** [reply]
> The seqcli binary is included in the Seq Windows installer

OK @nblumhardt almost there thank you for your continued assistance.  I see now I do have that on my development machine and I understand that command now.

> The main goal is to pipe Serilog's SelfLog to somewhere you can collect it on your application side.

Alright, I will consider this more then.  AppService is pretty locked down from what I can tell and they do not like using the local file system except for `Temp` directory... Even then it's not available to Kudu and is segregated away from browsing it.  If `Console.Error` writes to the Application Event Log there may be hope there as I am able to view eventlog.xml from Kudu with exception events.

**Comment by Mike-E-angelo** [comment, marked as answer]
Wanted to say that it has been forever since I have seen this.  This might have been a problem with .NET components and/or Azure infrastructure.  Whatever the reason it appears to have worked itself out.  Thank you again for all your assistance and care in building a great product. 🙏
